split-llvm-extract.cpp

```diff
@@ -59,42 +62,82 @@ llvm::cl::opt<bool> dry("d",
   std::unordered_set<llvm::GlobalVariable *> globals;
 
-  void visitInstruction(llvm::Instruction &instruction) {
-    std::stack<llvm::Value *> operands;
+  for (uint32_t i = 0; i < user.getNumOperands(); i++) {
+    operands.push(user.getOperand(i));
+  }
+
+  while (operands.size() > 0) {
+    auto currentOperand = operands.top();
+    operands.pop();
+
+    if (currentOperand == nullptr) {
+      // TODO: ignore null's for now.
+    } else if (auto globalVariable =
+                   llvm::dyn_cast<llvm::GlobalVariable>(currentOperand)) {
+      if (globalVariable->isConstant()) {
+        globals.insert(globalVariable);
+      } 
+    } else if (auto user = llvm::dyn_cast<llvm::User>(currentOperand)) {
+      for (uint32_t i = 0; i < user->getNumOperands(); i++) {
+        auto nextOperand = user->getOperand(i);
+        if (auto instruction =
+                llvm::dyn_cast<llvm::Instruction>(nextOperand)) {
+          continue;
+        } else {
+          operands.push(nextOperand);
+        }
+      }
+    }
+  }
+  return globals;
+}
 
-    for (uint32_t i = 0; i < instruction.getNumOperands(); i++) {
-      operands.push(instruction.getOperand(i));
+struct MyPass : public llvm::InstVisitor<MyPass> {
+  std::unordered_set<llvm::GlobalVariable *> globals;
+  // TODO: Move this to a separate function because globals 
+  // also need the full dependency tree moved.
+  void visitInstruction(llvm::Instruction &instruction) {
+    std::queue<llvm::GlobalVariable *> queueVariablesNeedMoving;
+    {
+      auto result = iterateOperands(instruction);
+      for(const auto item : result) {
+        queueVariablesNeedMoving.push(item);
+      }
     }
+    while(queueVariablesNeedMoving.size() > 0) {
+      auto currentVariable = queueVariablesNeedMoving.front();
+      queueVariablesNeedMoving.pop();
+      globals.insert(currentVariable);
 
-    while (operands.size() > 0) {
-      auto currentOperand = operands.top();
-      operands.pop();
-
-      if (currentOperand == nullptr) {
-        // TODO: ignore null's for now.
-      } else if (auto globalVariable =
-                     llvm::dyn_cast<llvm::GlobalVariable>(currentOperand)) {
-        if (globalVariable->isConstant()) {
-          globals.insert(globalVariable);
-        } 
-      } else if (auto user = llvm::dyn_cast<llvm::User>(currentOperand)) {
-        for (uint32_t i = 0; i < user->getNumOperands(); i++) {
-          auto nextOperand = user->getOperand(i);
-          if (auto instruction =
-                  llvm::dyn_cast<llvm::Instruction>(nextOperand)) {
-            continue;
-          } else {
-            operands.push(nextOperand);
+      if(currentVariable->hasInitializer()) {
+        auto alsoNeedMvoing = iterateOperands(*currentVariable->getInitializer());
+        for(const auto item : alsoNeedMvoing) {
+          if(globals.find(item) == globals.end()) {
+            queueVariablesNeedMoving.push(item);
           }
         }
-      }
+     }
```

Comment:
I think the removed space makes this line's indentation incorrect?

Reply:
Can I just do a llvm-format before merging?